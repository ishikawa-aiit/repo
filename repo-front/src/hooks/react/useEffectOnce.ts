import { useEffect, useRef, useState } from 'react'

/**
 * 依存関係がゼロの時、useEffect（）の代わりに使用するカスタムフック
 *
 * @remarks
 * このカスタムフックはReact 18の仕様変更のため、作成しています（React 18では依存関係がゼロの時、2回レンダリングされる）
 * https://github.com/vercel/next.js/issues/35822
 */
export const useEffectOnce = (effect: () => void | (() => void)) => {
  const destroyFunc = useRef<void | (() => void)>()
  const effectCalled = useRef(false)
  const renderAfterCalled = useRef(false)
  const [_val, setVal] = useState<number>(0)

  if (effectCalled.current) {
    renderAfterCalled.current = true
  }

  useEffect(() => {
    if (!effectCalled.current) {
      destroyFunc.current = effect()
      effectCalled.current = true
    }

    setVal((val) => {
      return val + 1
    })

    return () => {
      if (!renderAfterCalled.current) {
        return
      }
      if (destroyFunc.current) {
        destroyFunc.current()
      }
    }
    // マウント時に1回だけ実行するため
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])
}
