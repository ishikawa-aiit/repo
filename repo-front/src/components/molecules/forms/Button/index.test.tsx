import { screen, render, within } from '@testing-library/react'

import { Button } from '.'

const buttonText = 'サンプルテキスト'

describe('props:children', () => {
  it('正常にレンダリングされている', () => {
    const { container } = render(<Button>{buttonText}</Button>)

    const { getByText } = within(screen.getByTestId('Button'))

    expect(getByText(buttonText)).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })
})

describe('props:isDisabled', () => {
  it('trueの場合、buttonがdisabledである', () => {
    const { container } = render(<Button isDisabled>{buttonText}</Button>)

    expect(screen.getByTestId('Button')).toBeDisabled()
    expect(container).toMatchSnapshot()
  })

  it('falseの場合、buttonがenabledである', () => {
    const { container } = render(<Button>{buttonText}</Button>)

    expect(screen.getByTestId('Button')).toBeEnabled()
    expect(container).toMatchSnapshot()
  })
})
