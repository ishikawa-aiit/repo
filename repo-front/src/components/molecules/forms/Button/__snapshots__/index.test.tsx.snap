// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`props:children 正常にレンダリングされている 1`] = `
.emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: relative;
  white-space: nowrap;
  vertical-align: middle;
  outline: 2px solid transparent;
  outline-offset: 2px;
}

<div>
  <button
    class="chakra-button emotion-0"
    data-cy="Button"
    data-testid="Button"
    type="button"
  >
    サンプルテキスト
  </button>
</div>
`;

exports[`props:isDisabled falseの場合、buttonがenabledである 1`] = `
.emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: relative;
  white-space: nowrap;
  vertical-align: middle;
  outline: 2px solid transparent;
  outline-offset: 2px;
}

<div>
  <button
    class="chakra-button emotion-0"
    data-cy="Button"
    data-testid="Button"
    type="button"
  >
    サンプルテキスト
  </button>
</div>
`;

exports[`props:isDisabled trueの場合、buttonがdisabledである 1`] = `
.emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: relative;
  white-space: nowrap;
  vertical-align: middle;
  outline: 2px solid transparent;
  outline-offset: 2px;
}

<div>
  <button
    class="chakra-button emotion-0"
    data-cy="Button"
    data-testid="Button"
    disabled=""
    type="button"
  >
    サンプルテキスト
  </button>
</div>
`;
